### PR TITLE
Change the FeatureGates implemantation

### PR DESCRIPTION
**Pull request by nunnatsa**
Change the FeatureGates implemantation to be strict and to add documentation to the CR for each feature gate.

The previose implementation used a `map[string]bool` to store the feature gates. That way, it was impossible to add documentation for each feature gate.

This PR changes the implementation so the feature gates are now stored in a struct, as a pointer to boolean (to keep them optional). The result is a documentable feature gates as well as strict naming of the feature gates, because their names are not just astring but a field name.

Since `HotplugVolumes` is not a valid json or yaml tag name (starts with upper case), the gate name was changed to `hotplugVolumes` (starts with lower case)

As part of this change, the HyperConverged CRD contains the documentation for the `hotplugVolumes` feature gate, that can be read directly from the CRD, or by calling `kubectl explain`:
```
$ kubectl explain HyperConverged.spec.featureGates
KIND:     HyperConverged
VERSION:  hco.kubevirt.io/v1beta1

RESOURCE: featureGates <Object>

DESCRIPTION:
     featureGates HyperConvergedFeatureGates contains a list of feature enabler
     flags. Setting a flag to `true` will enable the feature. Setting `false` or
     removing the feature gate, disables the feature.

FIELDS:
   hotplugVolumes	<boolean>
     Allow attaching a data volume to a running VMI; type is boolean; default is
     false
```

In addition, this PR adds feature gates documentation to the `docs/cluster-configuration.md` file with the following usage example:

```yaml
apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
  name: kubevirt-hyperconverged
spec:
  infra: {}
  workloads: {}
  featureGates:
    hotplugVolumes: true
```

Signed-off-by: Nahshon Unna-Tsameret <nunnatsa@redhat.com>

**Release note**:
```release-note
Added CRD documentation to the hotplugVolumes feature gate
Added Feature gates documentation to the cluster-configuration.md document
Change HotplugVolumes to hotplugVolumes
```

### CODE REVIEW

File: deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/hco00.crd.yaml [hunk 1]
```diff
@@ -151,9 +151,11 @@ spec:
                 description: BareMetalPlatform indicates whether the infrastructure is baremetal.
                 type: boolean
               featureGates:
-                additionalProperties:
-                  type: boolean
-                description: featureGates HyperConvergedFeatureGates contain a list of feature enabler flags. Setting a flag to `true` will enable the feature. Setting `false` or removing the feature gate, disables the feature. optional+
+                description: featureGates HyperConvergedFeatureGates contains a list of feature enabler flags. Setting a flag to `true` will enable the feature. Setting `false` or removing the feature gate, disables the feature.
```
Comment (dankenigsberg) [inline deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/hco00.crd.yaml:154]: What is `HyperConvergedFeatureGates` ? Is it something that a typical admin understands (I don't).
I don't believe it is a list, do I think that s better description would be "featureGates is a map of feature enabler flags.

Is it no longer optional? Do I have to add it to all RC?

File: deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/hco00.crd.yaml [hunk 2]
```diff
@@ -151,9 +151,11 @@ spec:
...
+                properties:
+                  HotplugVolumes:
+                    description: Allow attaching a data volume to a running VMI; type is boolean; default is false
```
Comment (dankenigsberg) [inline deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/hco00.crd.yaml:157]: Do we really need to put the type in the English text? There's the `type` section for that. I would also not document the default. A missing flags can means that the admin does not care, and trusts HCO to do the right thing; we may want to flip the default without much notice.

File: docs/cluster-configuration.md [hunk 3]
```diff
@@ -109,3 +109,29 @@ tolerations.
           value: "kubevirt"
           effect: "NoSchedule"
   ```
+
+
+
+
+## Additional Configurations
```
Comment (dankenigsberg) [inline docs/cluster-configuration.md:116]: I find this section superfluous, adding very little info.. I'd promote featureGate section to level two (##)

**Comment by zcahana** [comment]
And another general comment - what is the significance (in terms of API design) of grouping various knobs under a `featureGates:` field? Is it to indicate the immaturity or the temporariness of the feature gate? Is it a "part" of the HyperConverged CR that we deliberately tell users that nested fields (e.g., `.hotplugVolumes`) will likely be deprecated soon and later removed (i.e. once the feature graduates, or withdrawn).

File: pkg/apis/hco/v1beta1/hyperconverged_types.go [hunk 4]
```diff
@@ -56,25 +57,31 @@ type HyperConvergedConfig struct {
...
-	fg, found := fgs[fgName]
-	return fg && found
+// HyperConvergedFeatureGates is a set of optional feature gates to enable or disable new features that are not enabled
+// by default yet.
+// +optional
+// +k8s:openapi-gen=true
+type HyperConvergedFeatureGates struct {
+	// Allow attaching a data volume to a running VMI; type is boolean; default is false
+	// +optional
+	HotplugVolumes *bool `json:"HotplugVolumes,omitempty"`
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:67]: Agree, however, this is the KubeVirt naming, and I prefer to use the exact casing as done there, to avoid confusion.

File: pkg/apis/hco/v1beta1/hyperconverged_types.go [hunk 5]
```diff
@@ -56,25 +57,31 @@ type HyperConvergedConfig struct {
...
 }
 
 // get list of feature gates from a specific operand list
-func (fgs HyperConvergedFeatureGates) GetFeatureGateList(candidates []string) []string {
-	res := make([]string, 0, len(fgs))
-	for _, fg := range candidates {
-		if fgs.IsEnabled(fg) {
-			res = append(res, fg)
-		}
+func (fgs *HyperConvergedFeatureGates) GetFeatureGateList() []string {
```
Comment (zcahana) [inline pkg/apis/hco/v1beta1/hyperconverged_types.go:71]: I personally don't think this function is that needed (see other comments), but if keeping it, and since it has a pointer receiver, maybe support the case in which `fgs == nil`, and return an empty slice (instead of panic as now).

File: pkg/apis/hco/v1beta1/hyperconverged_types.go [hunk 6]
```diff
@@ -56,25 +57,31 @@ type HyperConvergedConfig struct {
...
 
-type HyperConvergedFeatureGates map[string]bool
-
-func (fgs HyperConvergedFeatureGates) IsEnabled(fgName string) bool {
-	fg, found := fgs[fgName]
-	return fg && found
+// HyperConvergedFeatureGates is a set of optional feature gates to enable or disable new features that are not enabled
+// by default yet.
+// +optional
+// +k8s:openapi-gen=true
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:63]: Yes, it ignored and is not added to the CRD documentation without it.

File: pkg/components/components.go [hunk 7]
```diff
@@ -766,6 +766,7 @@ func GetOperatorCRD(relPath string) *extv1.CustomResourceDefinition {
 	for _, p := range pkgs {
 		err = packageErrors(p, packages.TypeError)
 		if err != nil {
+			fmt.Printf("%#v\n", err)
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/components/components.go:769]: Added for debug. Left here because it helped me to understand what went wrong. I find it usable.

File: pkg/apis/hco/v1beta1/hyperconverged_types.go [hunk 8]
```diff
@@ -56,25 +57,31 @@ type HyperConvergedConfig struct {
...
+	res := make([]string, 0, 1)
+
+	if fgs.IsHotplugVolumesEnabled() {
+		res = append(res, virtconfig.HotplugVolumesGate)
 	}
 
 	return res
 }
 
+func (fgs HyperConvergedFeatureGates) IsHotplugVolumesEnabled() bool {
```
Comment (zcahana) [inline pkg/apis/hco/v1beta1/hyperconverged_types.go:81]: Here too, it may be more usable to have a pointer receiver (since `Spec.FeatureGates` could be `nil`), and support the case in which `fgs == nil`. This will make the calls a bit simpler, without needing to check if `Spec.FeaturesGates != nil`.

File: pkg/controller/operands/kubevirt.go [hunk 9]
```diff
@@ -115,13 +106,15 @@ func NewKubeVirt(hc *hcov1beta1.HyperConverged, opts ...string) (*kubevirtv1.Kub
 		Workloads:         hcoConfig2KvConfig(hc.Spec.Workloads),
 	}
 
-	fgs := hc.Spec.FeatureGates.GetFeatureGateList(managedKvFeatureGates)
-	if len(fgs) > 0 {
-		if spec.Configuration.DeveloperConfiguration == nil {
-			spec.Configuration.DeveloperConfiguration = &kubevirtv1.DeveloperConfiguration{}
-		}
+	if hc.Spec.FeatureGates != nil {
+		fgs := hc.Spec.FeatureGates.GetFeatureGateList()
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/controller/operands/kubevirt.go:110]: Currentlly there are only KubeVirt gates, and they are collected by name. Each operand handler will take what relevant to its operand.

File: pkg/controller/operands/kubevirt.go [hunk 10]
```diff
@@ -25,16 +25,7 @@ const (
...
-	cmFeatureGates   = "DataVolumes,SRIOV,LiveMigration,CPUManager,CPUNodeDiscovery,Sidecar,Snapshot"
-)
-
-var (
-	// managedKvFeatureGates - list of KV feature gates that can be set/clear by adding/remove them
-	// from HyperConverged CR
-	managedKvFeatureGates = []string{
-		kvHotplugVolumes,
-	}
+	cmFeatureGates = "DataVolumes,SRIOV,LiveMigration,CPUManager,CPUNodeDiscovery,Sidecar,Snapshot"
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/controller/operands/kubevirt.go:28]: For now, I would keep it. This is going to be changed soon, when we'll drop the configmap for KV. I would not invest there more than what we need.

File: pkg/apis/hco/v1beta1/hyperconverged_types.go (hunk 5 above)
Comment (nunnatsa) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:71]: Needed for the configmap. butyes, we can live without it.

File: pkg/apis/hco/v1beta1/hyperconverged_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zcahana) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:63]: Right, just realized this ends up in `zz_generated.openapi.go`.

File: pkg/apis/hco/v1beta1/hyperconverged_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zcahana) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:67]: I don't think it'll cause more confusion than non-standard/non-conventional casing in a Kubernetes object schema.
IMO, adhering to the conventions here is of further importance than aligning with an ad-hoc format that is anyway supposed to be removed soon (the configmap), isn't it?
+ we have documentation and schema validation so that users can find their way to the light.

File: pkg/controller/operands/kubevirt.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (zcahana) [reply pkg/controller/operands/kubevirt.go]: The point I'm trying to make is, that when we'll add feature gates to another operand, it may break this code that assumes that "there are only KubeVirt gates". The assumption that the developer who'll add the next feature gate will know/remember to come back to `operands/kubevirt.go` to fix this is optimistic, but not necessarily holds.

[the remark this answers is not in the document]
Comment (nunnatsa) [reply pkg/controller/operands/kubevirt.go]: Let me try to explain again: this code only run in kv handler. Other handlers will have their own implementation. I don't think that code assumes that the fgs are only for kv.

[the remark this answers is not in the document]
Comment (zcahana) [reply pkg/controller/operands/kubevirt.go]: If, say, CDI adds a feature gate, won't it go under `HCO.Spec.FeatureGates.<CDI-Feature-Name>` and then returned from `GetFeaturedGateList()`, and transferred to virt-handler?

**Comment by enp0s3** [comment]
> And another general comment - what is the significance (in terms of API design) of grouping various knobs under a `featureGates:` field? Is it to indicate the immaturity or the temporariness of the feature gate? Is it a "part" of the HyperConverged CR that we deliberately tell users that nested fields (e.g., `.hotplugVolumes`) will likely be deprecated soon and later removed (i.e. once the feature graduates, or withdrawn).
@zcahana Yes. Once a preview/optional feature will grown up into a fully fledged one, we will remove it from the feature gate section and perhaps it will have it's own place in the HCO CR.

**Review by enp0s3 — APPROVED** [review]
/lgtm
@nunnatsa I am feeling uncomforatble with using `fmt.Print` in the code, but if there is a promise to refactor the code to add a logger then let's go forward with this code change.

**Comment by nunnatsa** [comment]
> @nunnatsa I am feeling uncomforatble with using `fmt.Print` in the code, but if there is a promise to refactor the code to add a logger then let's go forward with this code change.

Done.

**Review by tiraboschi — APPROVED** [review]

**Review by enp0s3 — APPROVED** [review]
/lgtm
